File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/effects/clip/ClipReferenceBox.swift
//
//  ClipReferenceBox.swift
//  StyleEngine/effects/clip — wave 46 (lane Y4).
//
//  css-masking-1 §7.1 REFERENCE BOX resolution for `clip-path` on iOS.
//  Every `<basic-shape>` (and the bare `<geometry-box>` form) resolves
//  against a box chosen by the `<geometry-box>` keyword: margin-box /
//  border-box (the default) / padding-box / content-box. The rect a
//  SwiftUI `Shape.path(in:)` receives from `.clipShape` is the BORDER box
//  — `.engineClipPath` sits inside `engineSpacingMargin` and outside the
//  padding / border paint (StyleBuilder.applyGroupEffects) — so the other
//  three derive from it with the element's own metrics. Byte-parallel twin
//  of Android's ClipReferenceBox.kt (which additionally has to find the
//  border box inside a margin-inflated node; iOS does not).
//  Measured on wave45-final: contentBox-1a clipped a 180px circle (border
//  box) where the ref has the 100px content-box circle; geometryBox-2's
//  `polygon(…) margin-box` kept only the roof; marginBox-1b/1c/1d's 200px
//  outline flooded 291px of green because the keyword form was a no-op.
//

import SwiftUI

// One resolved reference box: its rect (clip-rect space) + corner curves.
struct ClipReferenceFrame: Equatable {
    let rect: CGRect
    let topLeft: BorderRadiusCorner
    let topRight: BorderRadiusCorner
    let bottomRight: BorderRadiusCorner
    let bottomLeft: BorderRadiusCorner
}

enum ClipReferenceBox {

    /// The reference box `box` of an element whose border box is `rect`.
    /// Padding / content boxes inset by the used border widths and the
    /// paddings (css-backgrounds-3 §5.1 inner radii = outer − width); the
    /// margin box outsets by the DECLARED margins with the css-shapes-1
    /// §4 corner rule. With `.none` metrics every box equals `rect` and
    /// carries square corners — byte-identical to the pre-wave-46 shapes.
    static func resolve(_ box: ClipGeometryBox, metrics m: ClipBoxMetrics,
                        in rect: CGRect) -> ClipReferenceFrame {
        let radii = borderRadii(m.radius, in: rect)
        switch box {
        case .borderBox:
            return ClipReferenceFrame(rect: rect, topLeft: radii[0], topRight: radii[1],
                                      bottomRight: radii[2], bottomLeft: radii[3])
        case .paddingBox:
            return inset(rect, radii, t: m.borderTop, r: m.borderRight,
                         b: m.borderBottom, l: m.borderLeft)
        case .contentBox:
            return inset(rect, radii,
                         t: m.borderTop + m.paddingTop, r: m.borderRight + m.paddingRight,
                         b: m.borderBottom + m.paddingBottom, l: m.borderLeft + m.paddingLeft)
        case .marginBox:
            let out = CGRect(x: rect.minX - m.marginLeft, y: rect.minY - m.marginTop,
                             width: rect.width + m.marginLeft + m.marginRight,
                             height: rect.height + m.marginTop + m.marginBottom)
            // Per corner: the horizontal axis takes that side's horizontal
            // margin, the vertical axis the vertical one.
            func corner(_ c: BorderRadiusCorner, _ mx: CGFloat, _ my: CGFloat) -> BorderRadiusCorner {
                BorderRadiusCorner(x: marginOutsetRadius(c.x, mx), y: marginOutsetRadius(c.y, my))
            }
            return ClipReferenceFrame(
                rect: out,
                topLeft: corner(radii[0], m.marginLeft, m.marginTop),
                topRight: corner(radii[1], m.marginRight, m.marginTop),
                bottomRight: corner(radii[2], m.marginRight, m.marginBottom),
                bottomLeft: corner(radii[3], m.marginLeft, m.marginBottom))
        }
    }

    /// css-shapes-1 §4 margin-box corner rule: a corner radius r outset by
    /// a margin m is `r + m` when m ≤ 0 or r/m ≥ 1, else
    /// `r + m·(1 + (r/m − 1)³)` — a square corner (r = 0) STAYS square
    /// while a large radius grows by the full margin (WPT marginBox-1c:
    /// 50 + 25 = 75 → full circle; marginBox-1d: 10 + 50·(1 + (0.2 − 1)³)
    /// = 34.4).
    static func marginOutsetRadius(_ r: CGFloat, _ m: CGFloat) -> CGFloat {
        if m <= 0 || r >= m { return max(0, r + m) }
        let d = r / m - 1
        return max(0, r + m * (1 + d * d * d))
    }

    /// Inner-box curves: each axis loses the band on its side, floored at 0.
    private static func inset(_ outer: CGRect, _ radii: [BorderRadiusCorner],
                              t: CGFloat, r: CGFloat, b: CGFloat, l: CGFloat) -> ClipReferenceFrame {
        let rect = CGRect(x: outer.minX + l, y: outer.minY + t,
                          width: max(0, outer.width - l - r),
                          height: max(0, outer.height - t - b))
        func shrink(_ c: BorderRadiusCorner, _ dx: CGFloat, _ dy: CGFloat) -> BorderRadiusCorner {
            BorderRadiusCorner(x: max(0, c.x - dx), y: max(0, c.y - dy))
        }
        return ClipReferenceFrame(
            rect: rect,
            topLeft: shrink(radii[0], l, t), topRight: shrink(radii[1], r, t),
            bottomRight: shrink(radii[2], r, b), bottomLeft: shrink(radii[3], l, b))
    }

    /// Border-box corner curves, order TL / TR / BR / BL: percent axes
    /// resolve against the border box (css-backgrounds-3 §5.1 via
    /// BorderRadiusCorner.resolved), then the §5.1 overlap rule scales ALL
    /// radii by the smallest `side / (sum of adjacent radii)` ratio below 1.
    private static func borderRadii(_ cfg: BorderRadiusConfig, in rect: CGRect) -> [BorderRadiusCorner] {
        let tl = cfg.topLeft.resolved(in: rect.size)
        let tr = cfg.topRight.resolved(in: rect.size)
        let br = cfg.bottomRight.resolved(in: rect.size)
        let bl = cfg.bottomLeft.resolved(in: rect.size)
        func ratio(_ extent: CGFloat, _ sum: CGFloat) -> CGFloat {
            (sum > extent && sum > 0) ? extent / sum : 1
        }
        let f = min(min(ratio(rect.width, tl.x + tr.x), ratio(rect.width, bl.x + br.x)),
                    min(ratio(rect.height, tl.y + bl.y), ratio(rect.height, tr.y + br.y)))
        func scaled(_ c: BorderRadiusCorner) -> BorderRadiusCorner {
            f < 1 ? BorderRadiusCorner(x: c.x * f, y: c.y * f) : c
        }
        return [scaled(tl), scaled(tr), scaled(br), scaled(bl)]
    }

    /// A rounded rectangle with independent elliptical corners — the clip
    /// region for the bare `<geometry-box>` form. Four quarter-ellipse
    /// arcs stitched to four edges as ONE closed contour; degenerates to
    /// `Path(rect)` when every corner is square. SwiftUI's y-down
    /// `addArc(clockwise: false)` sweeps with INCREASING angle (visually
    /// clockwise), so each corner runs from its incoming edge to its
    /// outgoing edge: TR −π/2→0, BR 0→π/2, BL π/2→π, TL π→3π/2.
    static func roundedPath(_ f: ClipReferenceFrame) -> Path {
        let r = f.rect
        if f.topLeft.isSquare && f.topRight.isSquare && f.bottomRight.isSquare && f.bottomLeft.isSquare {
            return Path(r)
        }
        var p = Path()
        // Corner helper: one quarter ellipse of radii (rx, ry) centred at c,
        // sweeping +π/2 from `from` (radians, y-down so −π/2 is "up"). Built
        // as a unit-circle arc under a non-uniform scale — CoreGraphics'
        // own Bézier arc approximation, no hand-rolled control points —
        // then its curve elements are appended to the CURRENT subpath
        // (the arc's own leading move is dropped) so the outline stays one
        // closed contour.
        func arc(_ c: CGPoint, _ rx: CGFloat, _ ry: CGFloat, from: Double) {
            guard rx > 0, ry > 0 else { return }
            var unit = Path()
            unit.addArc(center: .zero, radius: 1, startAngle: .radians(from),
                        endAngle: .radians(from + .pi / 2), clockwise: false)
            let t = CGAffineTransform(translationX: c.x, y: c.y).scaledBy(x: rx, y: ry)
            unit.applying(t).forEach { element in
                switch element {
                case .move: break
                case .line(let to): p.addLine(to: to)
                case .quadCurve(let to, let control): p.addQuadCurve(to: to, control: control)
                case .curve(let to, let c1, let c2): p.addCurve(to: to, control1: c1, control2: c2)
                case .closeSubpath: break
                }
            }
        }
        p.move(to: CGPoint(x: r.minX + f.topLeft.x, y: r.minY))
        p.addLine(to: CGPoint(x: r.maxX - f.topRight.x, y: r.minY))
        arc(CGPoint(x: r.maxX - f.topRight.x, y: r.minY + f.topRight.y), f.topRight.x, f.topRight.y, from: -.pi / 2)
        p.addLine(to: CGPoint(x: r.maxX, y: r.maxY - f.bottomRight.y))
        arc(CGPoint(x: r.maxX - f.bottomRight.x, y: r.maxY - f.bottomRight.y), f.bottomRight.x, f.bottomRight.y, from: 0)
        p.addLine(to: CGPoint(x: r.minX + f.bottomLeft.x, y: r.maxY))
        arc(CGPoint(x: r.minX + f.bottomLeft.x, y: r.maxY - f.bottomLeft.y), f.bottomLeft.x, f.bottomLeft.y, from: .pi / 2)
        p.addLine(to: CGPoint(x: r.minX, y: r.minY + f.topLeft.y))
        arc(CGPoint(x: r.minX + f.topLeft.x, y: r.minY + f.topLeft.y), f.topLeft.x, f.topLeft.y, from: .pi)
        p.closeSubpath()
        return p
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/effects/clip/ClipReferenceBox.kt
package com.styleconverter.runtime.effects.clip

// Wave 46 (lane Y4) — css-masking-1 §5.1 REFERENCE BOX resolution for
// `clip-path` on Android.
//
// The spec resolves every `<basic-shape>` (and the bare `<geometry-box>`
// form) against a reference box chosen by the `<geometry-box>` keyword:
// margin-box / border-box (the default) / padding-box / content-box. The
// runtime's clip node (StyleApplier step 3, EffectsFacade.apply) is chained
// OUTSIDE the margin step (step 4, MarginApplier → `absolutePadding`) and
// OUTSIDE the position step (PositionApplier → `absoluteOffset`), so the
// `size` a Shape receives in createOutline is the MARGIN-band-inflated node
// and the painted border box sits inside it at (appliedLeft + offsetX,
// appliedTop + offsetY). Measured on wave45-final: WPT clip-path-ellipse-006
// (`margin: 50px; clip-path: ellipse()`) drew rx against a 250px-wide node
// instead of the 150px border box, and every `{geometry-box, shape}` wire
// rendered unclipped. This file turns the node rect into the four CSS boxes
// (plus their corner curves) so the shape factories in ClipPathApplier can
// resolve against the right one — and against the node rect unchanged
// (byte-identical arithmetic) when the element has no margin / offset.

import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.DpOffset
import androidx.compose.ui.unit.dp
import com.styleconverter.runtime.borders.radius.BorderRadiusConfig
import com.styleconverter.runtime.spacing.CollapsedMargin
import com.styleconverter.runtime.spacing.MarginApplier
import com.styleconverter.runtime.spacing.MarginConfig
import com.styleconverter.runtime.spacing.MarginValue
import com.styleconverter.runtime.spacing.SpacingContext
import com.styleconverter.runtime.spacing.resolveToDp
import kotlin.math.max
import kotlin.math.min

/** Four physical Dp bands — one value per side, top/right/bottom/left. */
data class ClipBands(val top: Dp, val right: Dp, val bottom: Dp, val left: Dp) {
    companion object {
        /** All-zero bands — no border / no padding / no margin. */
        val ZERO = ClipBands(0.dp, 0.dp, 0.dp, 0.dp)
    }
}

/**
 * The element's own box metrics, read from the same IR the layout steps
 * consume (ClipPathExtractor.extractBoxGeometry), so the reference box is
 * derived from the very numbers that sized the node.
 *
 * @param margin the element's margin config, null when it declared none.
 *   Kept as a CONFIG (not resolved bands) because the APPLIED block-axis
 *   bands depend on the parent's CSS2 §8.3.1 collapse override, which only
 *   the composition knows (BlockMarginCollapse.LocalCollapsedMargin — the
 *   harness strips a root's vertical margins through that same channel);
 *   [nodeInsets] resolves it with the override the margin step itself used.
 * @param borderWidths USED border widths (0 for a side whose style is
 *   none/hidden — CSS2 §8.5.3), the padding-box inset from the border box.
 * @param paddings resolved padding bands, the content-box inset from the
 *   padding box.
 * @param radius the element's border-radius config — the border box's
 *   corner curves, from which the other boxes' curves derive.
 * @param positionOffset the resolved position inset (PositionApplier
 *   .resolvedOffset) — the paint slides by it INSIDE the clip node, so the
 *   reference box slides with it.
 */
data class ClipBoxGeometry(
    val margin: MarginConfig? = null,
    val borderWidths: ClipBands = ClipBands.ZERO,
    val paddings: ClipBands = ClipBands.ZERO,
    val radius: BorderRadiusConfig = BorderRadiusConfig(),
    val positionOffset: DpOffset = DpOffset.Zero,
) {
    /** True when the border box can differ from the node rect at all. */
    val displacesNode: Boolean
        get() = margin != null || positionOffset != DpOffset.Zero

    companion object {
        /** No metrics: every reference box coincides with the node rect. */
        val NONE = ClipBoxGeometry()
    }
}

/**
 * Where the BORDER BOX sits inside the clip node: the applied positive
 * margin bands (the node's own `absolutePadding`), plus the translation the
 * paint undergoes inside the node — the position offset and the negative-
 * margin remainder MarginApplier chains as `Modifier.offset`.
 */
data class ClipNodeInsets(val bands: ClipBands, val shiftX: Dp, val shiftY: Dp) {
    companion object {
        /** Border box == node rect. */
        val NONE = ClipNodeInsets(ClipBands.ZERO, 0.dp, 0.dp)
    }
}

/** One resolved reference box: its rect (node space, px) + corner curves. */
data class ClipReferenceFrame(
    val rect: Rect,
    val topLeft: CornerRadius,
    val topRight: CornerRadius,
    val bottomRight: CornerRadius,
    val bottomLeft: CornerRadius,
)

object ClipReferenceBox {

    /**
     * Resolve [ClipNodeInsets] for a box, given the CSS2 §8.3.1 collapse
     * override the margin step received (null outside a collapsing flow).
     * Shares MarginApplier.resolvedInsets for the positive bands — literally
     * the numbers the margin step padded the node with — and repeats its
     * two-line negative-remainder arithmetic for the `Modifier.offset` half.
     * KNOWN LIMITS (stated, not hidden): the default SpacingContext (a
     * preserved em/ch/lh margin could resolve on a different basis than the
     * layout step's threaded context — no corpus clip-path carries one), and
     * `margin: auto` centering (wrapContentWidth moves the box inside a
     * wider node; auto reads as 0 here, so the clip stays node-anchored).
     */
    fun nodeInsets(geometry: ClipBoxGeometry, collapsed: CollapsedMargin?): ClipNodeInsets {
        val offset = geometry.positionOffset
        val margin = geometry.margin
            ?: return ClipNodeInsets(ClipBands.ZERO, offset.x, offset.y)
        val applied = MarginApplier.resolvedInsets(margin, collapsed = collapsed)
        // Negative remainder — MarginApplier.applyResolved: the positive part
        // of each side is padding, the rest translates the node's content.
        val r = margin.resolve(isRtl = false)
        val ctx = SpacingContext()
        fun len(v: MarginValue?): Dp = (v as? MarginValue.Length)?.let { resolveToDp(it.value, ctx) } ?: 0.dp
        val negX = (len(r.left) - applied.left) - (len(r.right) - applied.right)
        val negY = (len(r.top) - applied.top) - (len(r.bottom) - applied.bottom)
        return ClipNodeInsets(
            ClipBands(applied.top, applied.right, applied.bottom, applied.left),
            shiftX = offset.x + negX, shiftY = offset.y + negY,
        )
    }

    /**
     * The reference box [box] of an element whose clip node measures [size].
     * Border box = node minus the applied bands, shifted; the other three
     * derive from it per css-masking-1 §5.1 / css-shapes-1 §4 (margin box
     * outset by the DECLARED margins — collapsing moves the node, not the
     * element's own margin box; padding/content boxes inset by the used
     * border widths and paddings).
     */
    fun resolve(
        geometry: ClipBoxGeometry, box: ClipGeometryBox, insets: ClipNodeInsets,
        size: Size, density: Density,
    ): ClipReferenceFrame = with(density) {
        val sx = insets.shiftX.toPx(); val sy = insets.shiftY.toPx()
        val border = Rect(
            left = insets.bands.left.toPx() + sx, top = insets.bands.top.toPx() + sy,
            right = size.width - insets.bands.right.toPx() + sx,
            bottom = size.height - insets.bands.bottom.toPx() + sy,
        )
        // Corner curves of the border box: css-backgrounds-3 §4.1 percent
        // axes against the border box, then the §5.1 overlap scaling.
        val radii = borderRadii(geometry.radius, border, density)
        val bw = geometry.borderWidths; val pd = geometry.paddings
        when (box) {
            ClipGeometryBox.BORDER_BOX -> frame(border, radii)
            ClipGeometryBox.PADDING_BOX ->
                insetFrame(border, radii, bw.top.toPx(), bw.right.toPx(), bw.bottom.toPx(), bw.left.toPx())
            ClipGeometryBox.CONTENT_BOX ->
                insetFrame(border, radii,
                    bw.top.toPx() + pd.top.toPx(), bw.right.toPx() + pd.right.toPx(),
                    bw.bottom.toPx() + pd.bottom.toPx(), bw.left.toPx() + pd.left.toPx())
            ClipGeometryBox.MARGIN_BOX -> {
                val r = geometry.margin?.resolve(isRtl = false)
                val ctx = SpacingContext()
                fun len(v: MarginValue?): Float =
                    (v as? MarginValue.Length)?.let { resolveToDp(it.value, ctx).toPx() } ?: 0f
                val mt = len(r?.top); val mr = len(r?.right); val mb = len(r?.bottom); val ml = len(r?.left)
                val rect = Rect(border.left - ml, border.top - mt, border.right + mr, border.bottom + mb)
                ClipReferenceFrame(
                    rect,
                    topLeft = CornerRadius(marginOutsetRadius(radii[0].x, ml), marginOutsetRadius(radii[0].y, mt)),
                    topRight = CornerRadius(marginOutsetRadius(radii[1].x, mr), marginOutsetRadius(radii[1].y, mt)),
                    bottomRight = CornerRadius(marginOutsetRadius(radii[2].x, mr), marginOutsetRadius(radii[2].y, mb)),
                    bottomLeft = CornerRadius(marginOutsetRadius(radii[3].x, ml), marginOutsetRadius(radii[3].y, mb)),
                )
            }
        }
    }

    /**
     * css-shapes-1 §4 margin-box corner rule: a corner radius r outset by a
     * margin m is `r + m` when m ≤ 0 or r/m ≥ 1, else `r + m·(1 + (r/m − 1)³)`
     * — so a square corner (r = 0) STAYS square (0 + m·(1 − 1) = 0) while a
     * large radius grows by the full margin (WPT marginBox-1c: 50 + 25 = 75
     * → a full circle; marginBox-1d: 10 + 50·(1 + (0.2 − 1)³) = 34.4).
     */
    fun marginOutsetRadius(r: Float, m: Float): Float {
        if (m <= 0f || r >= m) return max(0f, r + m)
        val ratio = r / m
        val d = ratio - 1f
        return max(0f, r + m * (1f + d * d * d))
    }

    /** Inner-box curves: css-backgrounds-3 §4.1 "inner radius = outer − width", floored at 0. */
    private fun insetFrame(outer: Rect, radii: Array<CornerRadius>, t: Float, r: Float, b: Float, l: Float) =
        ClipReferenceFrame(
            Rect(outer.left + l, outer.top + t, max(outer.left + l, outer.right - r), max(outer.top + t, outer.bottom - b)),
            topLeft = CornerRadius(max(0f, radii[0].x - l), max(0f, radii[0].y - t)),
            topRight = CornerRadius(max(0f, radii[1].x - r), max(0f, radii[1].y - t)),
            bottomRight = CornerRadius(max(0f, radii[2].x - r), max(0f, radii[2].y - b)),
            bottomLeft = CornerRadius(max(0f, radii[3].x - l), max(0f, radii[3].y - b)),
        )

    private fun frame(rect: Rect, radii: Array<CornerRadius>) =
        ClipReferenceFrame(rect, radii[0], radii[1], radii[2], radii[3])

    /**
     * Border-box corner curves in px, order TL / TR / BR / BL. A percent
     * axis (fraction wins over the Dp placeholder, as BorderRadiusConfig
     * documents) resolves against the border box's width (x) / height (y);
     * then css-backgrounds-3 §4.5's overlap rule scales ALL radii by the
     * smallest `side / (sum of the two adjacent radii)` ratio below 1.
     */
    private fun borderRadii(cfg: BorderRadiusConfig, box: Rect, density: Density): Array<CornerRadius> {
        fun axis(dp: Dp, fraction: Float?, extent: Float): Float =
            fraction?.let { extent * it } ?: with(density) { dp.toPx() }
        fun corner(pair: Pair<Dp, Dp>, fr: Pair<Float?, Float?>) =
            CornerRadius(axis(pair.first, fr.first, box.width), axis(pair.second, fr.second, box.height))
        val tl = corner(cfg.topStart, cfg.topStartFraction)
        val tr = corner(cfg.topEnd, cfg.topEndFraction)
        val br = corner(cfg.bottomEnd, cfg.bottomEndFraction)
        val bl = corner(cfg.bottomStart, cfg.bottomStartFraction)
        fun ratio(extent: Float, sum: Float) = if (sum > extent && sum > 0f) extent / sum else 1f
        val f = min(
            min(ratio(box.width, tl.x + tr.x), ratio(box.width, bl.x + br.x)),
            min(ratio(box.height, tl.y + bl.y), ratio(box.height, tr.y + br.y)),
        )
        fun scaled(c: CornerRadius) = if (f < 1f) CornerRadius(c.x * f, c.y * f) else c
        return arrayOf(scaled(tl), scaled(tr), scaled(br), scaled(bl))
    }
}
